vm/rust/src/lib.rs

```diff
@@ -98,6 +111,106 @@ pub extern "C" fn cairoVMCall(
     }
 }
 
+#[no_mangle]
+pub extern "C" fn cairoVMExecute(
+    txns_json: *const c_char,
+    classes_json: *const c_char,
+    reader_handle: usize,
+    block_number: c_ulonglong,
+    block_timestamp: c_ulonglong,
+    chain_id: *const c_char,
+    sequencer_address: *const c_uchar,
+) {
+    let reader = JunoStateReader::new(reader_handle);
+    let chain_id_str = unsafe { CStr::from_ptr(chain_id) }.to_str().unwrap();
+    let txn_json_str = unsafe { CStr::from_ptr(txns_json) }.to_str().unwrap();
+    let sn_api_txns: Result<Vec<StarknetApiTransaction>, serde_json::Error> =
+        serde_json::from_str(txn_json_str);
+    if sn_api_txns.is_err() {
+        report_error(reader_handle, sn_api_txns.unwrap_err().to_string().as_str());
+        return;
+    }
+
+    let mut classes: Result<Vec<Box<serde_json::value::RawValue>>, serde_json::Error> = Ok(vec![]);
+    if !classes_json.is_null() {
+        let classes_json_str = unsafe { CStr::from_ptr(classes_json) }.to_str().unwrap();
+        classes = serde_json::from_str(classes_json_str);
+    }
+    if classes.is_err() {
+        report_error(reader_handle, classes.unwrap_err().to_string().as_str());
+        return;
+    }
+
+    let sn_api_txns = sn_api_txns.unwrap();
+    let mut classes = classes.unwrap();
+
+    let sequencer_address_felt = ptr_to_felt(sequencer_address);
+    let block_context: BlockContext = build_block_context(
+        chain_id_str,
+        block_number,
+        block_timestamp,
+        sequencer_address_felt,
+    );
+    let mut state = CachedState::new(reader);
+
+    for sn_api_txn in sn_api_txns {
+        let contract_class = match sn_api_txn.clone() {
+            StarknetApiTransaction::Declare(declare) => {
+                if classes.len() == 0 {
+                    report_error(reader_handle, "missing declared class".to_string().as_str());
+                    return;
+                }
+                let class_json_str = classes.remove(0);
+
+                let mut maybe_cc = contract_class_from_json_str(class_json_str.get());
+                if declare.version() == TransactionVersion(2u32.into()) && maybe_cc.is_err() {
+                    // class json could be sierra
+                    maybe_cc = contract_class_from_sierra_json(class_json_str.get())
+                };
```

Comment:
Can you please explain why both `contract_class_from_json_str` and `contract_class_from_sierra_json` are required?

AFAIU both function above return `ContractClass` while `contract_class_from_json_str` can return either `ContractClassV0/V1` and we set the class json for `v1` to be cairo0 json while `v2` sets seirra json, so is `let v1_class = ContractClassV1::try_from_json_string(raw_json);` (line in `contract_class_from_json_str`) not sufficient to created `ContractClassV1`?

If not then we should not try to create `ContractClassV1` in `contract_class_from_json_str`

Reply:
`contract_class_from_json_str` deals with compiled classes only. It is used in the JunoStateReader as well.

a declared class may be a legacy casm class (ie ContractClassV0), or a Sierra class that we first need to compile to a casm class (ie ContractClassV1)

for the first case, a just reused `contract_class_from_json_str` and the second case is implemented in `contract_class_from_sierra_json`. Which parses the sierra class and compiles it to a `ContractClassV1`